lib/at_cmd/at_cmd.c

```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
 
 static void callback_worker(struct k_work *item)
 {
-	struct callback_work_item *data =
-		CONTAINER_OF(item, struct callback_work_item, work);
+	ARG_UNUSED(item);
+	struct resp_item resp;
 
-	if (data != NULL) {
-		data->callback(data->data);
+	while (k_msgq_get(&responses, &resp, K_NO_WAIT) == 0) {
+		if (resp.callback != NULL ||
+		    resp.buf != NULL) {
+			resp.callback(resp.buf);
+		}
+		if (resp.flags & AT_RESP_ALLOC) {
+			k_free(resp.buf);
+		}
 	}
+}
+
+
 
-	k_mem_slab_free(&rsp_work_items, (void **)&data);
+/* Distributes an AT response to the appropriate queue */
+static void distribute_resp(const struct cmd_item *cmd, struct resp_item *resp)
+{
+	if (resp->state != AT_CMD_NOTIFICATION &&
+	    cmd->flags & AT_CMD_SYNC) {
+		LOG_DBG("Enqueueing response for sync call");
+		k_msgq_put(&response_sync, resp, K_FOREVER);
+	} else {
+		LOG_DBG("Enqueueing response for callback queue");
+		k_msgq_put(&responses, resp, K_FOREVER);
+		k_work_submit(&callback_work);
+	}
 }
 
+static inline int at_write(const char *const cmd)
+{
+	int bytes_sent;
+	int bytes_to_send = strlen(cmd);
+
+	LOG_DBG("Sending command %s", log_strdup(cmd));
+
+	bytes_sent = send(common_socket_fd, cmd, bytes_to_send, 0);
+
+	if (bytes_sent == -1) {
+		LOG_ERR("Failed to send AT command (err:%d)", errno);
+		return -errno;
+	}
+
+	if (bytes_sent != bytes_to_send) {
+		LOG_ERR("Bytes sent (%d) was not the "
+			"same as expected (%d)",
+			bytes_sent, bytes_to_send);
```

Comment:
Missing return.
But I think you can make this simpler. Just print and return the same error for both.

Reply:
The lack of return here is inherited from the current implementation (64bdc76384a31cc60d10aeb9a37795cdd2175ed2) and, I think, intentional. If a partial AT command is in fact sent, at_cmd has to wait for the (error) response. It seems like an unnecessary check, if `send` either fails or sends the correct length. @evenl  , do you know if it is still needed?

Reply:
Handled offline